### PR TITLE
Behavioral change: strip IPv6 zone identifiers during lookup

### PR DESCRIPTION
**Pull request by gaissmai**
Unconditionally strip zone identifiers from IPv6 addresses in leaf nodes during Contains and Lookup operations.

Previously, passing a zoned IPv6 address resulted in undefined behavior (e.g. false negatives). Zoned IPv6 addresses are now safely supported and the zone identifier is ignored during lookup.

Update docstrings accordingly.

## Summary by CodeRabbit

* **New Features**
  * IPv6 zone identifiers are ignored during address containment and lookup operations, producing the same results as unzoned addresses.
  * Invalid addresses return unsuccessful results.

* **Breaking Changes**
  * Bitset operation names and behavior have changed; integrations using the previous methods may require updates.

* **Tests**
  * Added coverage for zoned IPv6 lookups, fallback matching, invalid probes, and benchmark scenarios.

### CODE REVIEW

File: commonmethods_tmpl.go [hunk 1]
```diff
@@ -193,6 +194,11 @@ LOOP:
 			return kid.Value, true
 
 		case *nodes.LeafNode[V]:
+			if !is4 {
+				// strip the zone unconditionally
+				// see https://github.com/gaissmai/bart/pull/418#issuecomment-5735613506
+				ip = netip.PrefixFrom(ip, 0).Addr()
+			}
 			if kid.Prefix.Contains(ip) {
```
[the remark this answers is not in the document]
Comment (gaissmai) [reply commonmethods_tmpl.go]: @greptileai kannst du mir so einen test schreiben?

**Comment by gaissmai** [comment]
@database64128 Could you do me a favor and review this PR regarding IPv6 zone stripping?

**Comment by database64128** [comment]
You mentioned earlier that you were going to benchmark it. Since you didn't post any benchmark results, I got curious and ran some rounds:

```console
goos: linux
goarch: amd64
pkg: github.com/gaissmai/bart
cpu: Intel(R) Core(TM) Ultra 9 285K
                                │ /tmp/bart-main.txt │         /tmp/bart-devel.txt         │         /tmp/bart-is6.txt          │       /tmp/bart-withzone.txt        │
                                │       sec/op       │   sec/op     vs base                │   sec/op     vs base               │   sec/op     vs base                │
FullFastMatch6/Contains-24               12.55n ± 3%   12.41n ± 1%        ~ (p=0.072 n=10)   12.79n ± 4%  +1.91% (p=0.043 n=10)   12.21n ± 1%   -2.71% (p=0.009 n=10)
FullFastMiss6/Contains-24                11.38n ± 1%   11.08n ± 3%   -2.64% (p=0.000 n=10)   11.12n ± 3%  -2.24% (p=0.000 n=10)   10.85n ± 1%   -4.66% (p=0.000 n=10)
FullBartMatch6/Contains-24               19.11n ± 1%   19.01n ± 1%        ~ (p=0.147 n=10)   18.83n ± 3%  -1.52% (p=0.003 n=10)   18.34n ± 1%   -4.05% (p=0.000 n=10)
FullBartMiss6/Contains-24                16.74n ± 1%   16.66n ± 2%        ~ (p=0.469 n=10)   16.58n ± 0%  -0.96% (p=0.019 n=10)   16.34n ± 1%   -2.36% (p=0.003 n=10)
BartWorstCaseMatch6/Contains-24          77.77n ± 1%   63.14n ± 2%  -18.81% (p=0.000 n=10)   76.84n ± 3%  -1.18% (p=0.001 n=10)   63.01n ± 0%  -18.97% (p=0.000 n=10)
FastWorstCaseMatch6/Contains-24          41.94n ± 0%   30.27n ± 0%  -27.82% (p=0.000 n=10)   42.46n ± 0%  +1.25% (p=0.000 n=10)   30.30n ± 1%  -27.73% (p=0.000 n=10)
BartWorstCaseMiss6/Contains-24           74.00n ± 0%   61.08n ± 0%  -17.46% (p=0.000 n=10)   72.84n ± 1%  -1.55% (p=0.000 n=10)   60.70n ± 1%  -17.96% (p=0.000 n=10)
FastWorstCaseMiss6/Contains-24           41.22n ± 1%   29.89n ± 2%  -27.50% (p=0.000 n=10)   41.03n ± 0%  -0.49% (p=0.000 n=10)   29.98n ± 1%  -27.29% (p=0.000 n=10)
geomean                                  28.66n        24.99n       -12.80%                  28.48n       -0.61%                  24.69n       -13.85%
```

- `is6`: `if !is4` ➡️ `if ip.Is6()`
- `withzone`: `ip = netip.PrefixFrom(ip, 0).Addr()` ➡️ `ip = ip.WithZone("")`

I don't understand how this change results in significant speedup in worst case benchmarks.

**Comment by gaissmai** [comment]
I conducted special benchmarks exclusively for the leaf nodes and observed no significant performance impact from zone stripping. I suspect there may be an anomaly or flaw in the worstcase benchmark suite. I'll dump the trie tomorrow and look at it.

**Comment by database64128** [comment]
And here's the result of worst case benchmarks with `worstCaseProbeIP6` changed to `"ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff%eno1"`:

```console
goos: linux
goarch: amd64
pkg: github.com/gaissmai/bart
cpu: Intel(R) Core(TM) Ultra 9 285K
                                │ /tmp/bart-main-z.txt │        /tmp/bart-devel-z.txt        │        /tmp/bart-is6-z.txt         │      /tmp/bart-withzone-z.txt       │
                                │        sec/op        │   sec/op     vs base                │   sec/op     vs base               │   sec/op     vs base                │
BartWorstCaseMatch6/Contains-24            77.63n ± 4%   61.99n ± 3%  -20.15% (p=0.000 n=10)   76.53n ± 3%  -1.42% (p=0.019 n=10)   62.32n ± 4%  -19.72% (p=0.000 n=10)
FastWorstCaseMatch6/Contains-24            41.63n ± 2%   29.55n ± 3%  -29.01% (p=0.000 n=10)   42.25n ± 1%  +1.48% (p=0.001 n=10)   29.77n ± 2%  -28.50% (p=0.000 n=10)
BartWorstCaseMiss6/Contains-24             73.41n ± 3%   60.26n ± 2%  -17.91% (p=0.000 n=10)   72.24n ± 3%       ~ (p=0.101 n=10)   59.41n ± 3%  -19.07% (p=0.000 n=10)
FastWorstCaseMiss6/Contains-24             40.93n ± 2%   29.40n ± 3%  -28.17% (p=0.000 n=10)   40.67n ± 2%       ~ (p=0.469 n=10)   29.20n ± 4%  -28.67% (p=0.000 n=10)
geomean                                    55.82n        42.44n       -23.96%                  55.51n       -0.55%                  42.35n       -24.13%
```

**Comment by gaissmai** [comment]
First, I ran comparative benchmarks between main and devel and see no regression whatsoever.

```text
benchstat main-1cdf7a8.bm devel-8e72b78.bm 
goos: linux
goarch: amd64
pkg: github.com/gaissmai/bart
cpu: AMD Ryzen 7 PRO 4750U with Radeon Graphics
                             │ main-1cdf7a8.bm │          devel-8e72b78.bm          │
                             │     sec/op      │   sec/op     vs base               │
BartWorstCaseMatch6/Contains       89.68n ± 0%   91.01n ± 1%  +1.47% (p=0.000 n=10)
FastWorstCaseMatch6/Contains       56.49n ± 1%   56.85n ± 1%  +0.65% (p=0.000 n=10)
BartWorstCaseMiss6/Contains        88.52n ± 1%   87.41n ± 1%  -1.26% (p=0.000 n=10)
FastWorstCaseMiss6/Contains        54.87n ± 0%   55.73n ± 0%  +1.57% (p=0.000 n=10)
geomean                            70.43n        70.85n       +0.60%
```

**Comment by gaissmai** [comment]
cu tomorrow, I live in the Berlin timezone ;)

**Comment by gaissmai** [comment]
maybe you can test your code variations with this test?

```go
 package bart

import (
	"net/netip"
	"testing"
)

var leafPfxs = []netip.Prefix{
	mpp("10.0.0.0/23"),
	mpp("2001:db8::/47"),
}

func TestMy(t *testing.T) {
	fast := new(Fast[string])
	for _, pfx := range leafPfxs {
		fast.Insert(pfx, pfx.String())
	}
	t.Log(fast.dumpString())
	t.Fail()
}

func BenchmarkMy(b *testing.B) {
	fast := new(Fast[string])
	for _, pfx := range leafPfxs {
		fast.Insert(pfx, pfx.String())
	}

	b.Run("Lookup IPv4", func(b *testing.B) {
		ips := []netip.Addr{
			mpa("10.0.0.0"),
			mpa("10.0.0.1"),
			mpa("10.0.0.2"),
			mpa("10.0.0.3"),
		}
		i := 0
		for b.Loop() {
			_, _ = fast.Lookup(ips[i&3])
			i++
		}
	})

	b.Run("Lookup IPv6zoned", func(b *testing.B) {
		ips := []netip.Addr{
			mpa("2001:db8:0:0::%eth0"),
			mpa("2001:db8:0:0::1%eth0"),
			mpa("2001:db8:0:0::2%eth0"),
			mpa("2001:db8:0:0::3%eth0"),
		}
		i := 0
		for b.Loop() {
			_, _ = fast.Lookup(ips[i&3])
			i++
		}
	})
}
```

**Comment by database64128** [comment]
Sure, here you go:

```console
goos: linux
goarch: amd64
pkg: github.com/gaissmai/bart
cpu: Intel(R) Core(TM) Ultra 9 285K
                       │ /tmp/bart-main-my.txt │       /tmp/bart-devel-my.txt       │        /tmp/bart-is6-my.txt        │      /tmp/bart-withzone-my.txt      │
                       │        sec/op         │   sec/op     vs base               │   sec/op     vs base               │   sec/op     vs base                │
My/Lookup_IPv4-24                  4.194n ± 2%   4.110n ± 3%  -2.01% (p=0.002 n=10)   4.200n ± 2%       ~ (p=0.956 n=10)   4.203n ± 2%        ~ (p=0.753 n=10)
My/Lookup_IPv6zoned-24             4.492n ± 3%   4.575n ± 4%       ~ (p=0.105 n=10)   4.646n ± 2%  +3.45% (p=0.000 n=10)   5.099n ± 2%  +13.53% (p=0.000 n=10)
geomean                            4.340n        4.336n       -0.10%                  4.417n       +1.77%                  4.629n        +6.66%
```

All test runs produced the same output:

```console
=== RUN   TestMy
    my_test.go:18: 
        ### IPv4: size(1), subnodes(1), prefixes(0), leaves(1), fringes(0)
        [STOP] depth:  0 path: [] / 0
        octets(#1): [10]
        leaves(#1): 10:{10.0.0.0/23, 10.0.0.0/23}
        
        ### IPv6: size(1), subnodes(1), prefixes(0), leaves(1), fringes(0)
        [STOP] depth:  0 path: [] / 0
        octets(#1): [32]
        leaves(#1): 0x20:{2001:db8::/47, 2001:db8::/47}
```

Running these benchmarks pushed the CPU temperature to ~75C sustained, and the CPU cooler fans were spinning at maximum RPM. Existing benchmarks also max out a single core, but the CPU temperature only swings between 55C and 65C, so it's much quieter.

**Comment by database64128** [comment]
Off topic nit: It's totally OK to write `for i := 0; b.Loop(); i++`. I actually commented on it when Go 1.24 came out: https://github.com/golang/go/issues/61515#issuecomment-2656656554. The Go team member who implemented `B.Loop` then sent a change to update the documentation: golang/go@f48b53f0f62c94fac8d835c8e1b48fab5b842bd3

**Review by database64128 — APPROVED** [review]
I don't know how to explain the benchmark results, but I don't think there's anything wrong with them. Actually, if there's anything we learned, it's that my "clever" trick doesn't actually improve performance, at least for large tables used by the benchmarks in the repository. For small tables like the one in `BenchmarkMy`, the slowdown is only ~11%:

```console
goos: linux
goarch: amd64
pkg: github.com/gaissmai/bart
cpu: Intel(R) Core(TM) Ultra 9 285K
                       │ /tmp/bart-devel-my.txt │      /tmp/bart-withzone-my.txt      │
                       │         sec/op         │   sec/op     vs base                │
My/Lookup_IPv4-24                   4.110n ± 3%   4.203n ± 2%   +2.26% (p=0.006 n=10)
My/Lookup_IPv6zoned-24              4.575n ± 4%   5.099n ± 2%  +11.47% (p=0.000 n=10)
geomean                             4.336n        4.629n        +6.77%
```

Not sure if this justifies the added complexity. But of course, this is your call.

Other than that, LGTM!

**Comment by gaissmai** [comment]
> And here's the result of worst case benchmarks with `worstCaseProbeIP6` changed to `"ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff%eno1"`:
> 
> ```
> goos: linux
> goarch: amd64
> pkg: github.com/gaissmai/bart
> cpu: Intel(R) Core(TM) Ultra 9 285K
>                                 │ /tmp/bart-main-z.txt │        /tmp/bart-devel-z.txt        │        /tmp/bart-is6-z.txt         │      /tmp/bart-withzone-z.txt       │
>                                 │        sec/op        │   sec/op     vs base                │   sec/op     vs base               │   sec/op     vs base                │
> BartWorstCaseMatch6/Contains-24            77.63n ± 4%   61.99n ± 3%  -20.15% (p=0.000 n=10)   76.53n ± 3%  -1.42% (p=0.019 n=10)   62.32n ± 4%  -19.72% (p=0.000 n=10)
> FastWorstCaseMatch6/Contains-24            41.63n ± 2%   29.55n ± 3%  -29.01% (p=0.000 n=10)   42.25n ± 1%  +1.48% (p=0.001 n=10)   29.77n ± 2%  -28.50% (p=0.000 n=10)
> BartWorstCaseMiss6/Contains-24             73.41n ± 3%   60.26n ± 2%  -17.91% (p=0.000 n=10)   72.24n ± 3%       ~ (p=0.101 n=10)   59.41n ± 3%  -19.07% (p=0.000 n=10)
> FastWorstCaseMiss6/Contains-24             40.93n ± 2%   29.40n ± 3%  -28.17% (p=0.000 n=10)   40.67n ± 2%       ~ (p=0.469 n=10)   29.20n ± 4%  -28.67% (p=0.000 n=10)
> geomean                
> ```

Re-verified that the worst-case benchmark `trie` does not use `leaf compression`; CPU profiling also shows this path is never hit, which matches my benchmark results:

```text
benchstat main-1cdf7a8-z.bm devel-6071dd5.bm devel-withzone.bm 
goos: linux
goarch: amd64
pkg: github.com/gaissmai/bart
cpu: AMD Ryzen 7 PRO 4750U with Radeon Graphics
                             │ main-1cdf7a8-z.bm │          devel-6071dd5.bm          │         devel-withzone.bm          │
                             │      sec/op       │   sec/op     vs base               │   sec/op     vs base               │
BartWorstCaseMatch6/Contains         89.21n ± 2%   88.65n ± 0%       ~ (p=0.078 n=10)   88.63n ± 1%       ~ (p=0.448 n=10)
FastWorstCaseMatch6/Contains         56.64n ± 1%   56.79n ± 0%  +0.27% (p=0.008 n=10)   56.88n ± 0%  +0.42% (p=0.001 n=10)
BartWorstCaseMiss6/Contains          88.79n ± 0%   85.62n ± 0%  -3.56% (p=0.000 n=10)   86.01n ± 0%  -3.13% (p=0.000 n=10)
FastWorstCaseMiss6/Contains          54.64n ± 1%   54.06n ± 0%  -1.05% (p=0.000 n=10)   54.31n ± 0%  -0.61% (p=0.000 n=10)
geomean                              70.36n        69.48n       -1.25%                  69.66n       -1.00%
```

**Comment by gaissmai** [comment]
@database64128
Now for the **single-leaf trie** tests. This is an artificial edge case, ~~but showing an expected negative impact on IPv4 as well, due to the added v4/v6 check in the hot path.~~

```text
benchstat main-1cdf7a8.bm devel-3496ecd.bm 
goos: linux
goarch: amd64
pkg: github.com/gaissmai/bart
cpu: AMD Ryzen 7 PRO 4750U with Radeon Graphics
                    │ main-1cdf7a8.bm │           devel-3496ecd.bm           │
                    │     sec/op      │    sec/op     vs base                │
My/Lookup_IPv4            9.809n ± 0%    9.637n ± 0%   -1.75% (p=0.000 n=10)
My/Lookup_IPv6zoned       9.213n ± 2%   10.390n ± 1%  +12.78% (p=0.000 n=10)
geomean                   9.506n         10.01n        +5.26%
```

No, after removing the non-inlinable method again, it makes no difference for IPv4, which is what made me suspicious in the first place.
The IPv6 case can't be compared with the `main` branch, of course, because `netip.Prefix.Contains` immediately returns false on the `main` branch since the IPv6 address now carries a zone identifier.

But I can demonstrate that your `clever trick` for inlining withoutZone actually makes a noticeable difference:

```text
benchstat devel-3496ecd.bm devel-withzone.bm 
goos: linux
goarch: amd64
pkg: github.com/gaissmai/bart
cpu: AMD Ryzen 7 PRO 4750U with Radeon Graphics
                    │ devel-3496ecd.bm │         devel-withzone.bm          │
                    │      sec/op      │   sec/op     vs base               │
My/Lookup_IPv4             9.637n ± 0%   9.709n ± 0%  +0.75% (p=0.000 n=10)
My/Lookup_IPv6zoned        10.39n ± 1%   11.40n ± 1%  +9.72% (p=0.000 n=10)
geomean                    10.01n        10.52n       +5.14%
```

edited: ~~Correct text,~~ but values are off due to non-inlinable functions. Pushing an update with fresh values coming soon.

edited-2: now with hopefully correct values
edited-3: now with correct values :(

**Comment by gaissmai** [comment]
@coderabbit review

**Comment by gaissmai** [comment]
@database64128 I plan to merge this tomorrow and publish it as release v0.30.0, provided no further concerns arise overnight.
